packages/flame/lib/src/game/mixins/component_tree_root.dart

```diff
@@ -0,0 +1,153 @@
+import 'dart:collection';
+
+import 'package:meta/meta.dart';
+import 'package:vector_math/vector_math_64.dart';
+
+import '../../components/component.dart';
+import 'game.dart';
+
+/// This mixin is used to designate a class as a root of the component tree.
+///
+/// There should be only one [ComponentTreeRoot] in an application. Usually this
+/// role is fulfilled by the `FlameGame`, so you rarely use this mixin directly.
+///
+/// The purpose of this mixin is to host common facilities for component
+/// lifecycle management. These include: lifecycle event queues, and the canvas
+/// size used during [Component.onGameResize].
+///
+/// In order to use this mixin in a custom [Game], do the following:
+///   - add this mixin to your game class;
+///   - set it as the default [Component.root];
+///   - call [processComponentQueues] on every game tick.
+mixin ComponentTreeRoot on Game {
+  /// Global queues for adding children to components.
+  ///
+  /// The keys in this map are the parents, and the values are the queues of
+  /// components that need to be added to the parents.
+  ///
+  /// When the user `add()`s a component to a parent, we immediately place it
+  /// into the parent's queue, and only after that do the standard lifecycle
+  /// processing: resizing, loading, mounting, etc. After all that is finished,
+  /// the component is retrieved from the queue and placed into the parent's
+  /// children list.
+  ///
+  /// Since the components are processed in the FIFO order, this ensures that
+  /// they will be added to the parent in exactly the same order as the user
+  /// invoked `add()`s, even though they are loading asynchronously and may
+  /// finish loading in arbitrary order.
+  final Map<Component, Queue<Component>> childrenQueue = {};
+
+  /// Global queue of components that need to be mounted.
+  ///
+  /// Occasionally, a component may be added to a parent that hasn't been
+  /// mounted yet. In such a case, the child component must wait until its
+  /// parent mounts before it can mount too. Such components are placed into
+  /// this queue.
+  final Queue<Component> mountQueue = Queue();
```

Comment:
When a parent is mounted, couldn't that just propagate to all it's children that their `onMount` methods should run?

Reply:
I'll need to think about that.

Reply:
Yes, I was able to remove this queue.

Reply:
Nice!